modules/3d/src/rgbd/hash_tsdf_functions.cpp

```diff
@@ -0,0 +1,1534 @@
+                    continue;
+                Point3f camPoint = reproj(Point3f((float)x, (float)y, z));
+                Point3f volPoint = cam2vol * camPoint;
+                //! Find accessed TSDF volume unit for valid 3D vertex
+                Vec3i lower_bound = volumeToVolumeUnitIdx(volPoint - truncPt, volumeUnitSize);
+                Vec3i upper_bound = volumeToVolumeUnitIdx(volPoint + truncPt, volumeUnitSize);
+
+                int pixLocalCounter = 0;
+                LocalVolUnits pixLocalVolUnits;
+                for (int i = lower_bound[0]; i <= upper_bound[0]; i++)
+                    for (int j = lower_bound[1]; j <= upper_bound[1]; j++)
+                        for (int k = lower_bound[2]; k <= upper_bound[2]; k++)
+                        {
+                            const Vec3i tsdf_idx = Vec3i(i, j, k);
+
+                            if (hashTable.find(tsdf_idx) < 0)
+                            {
+                                bool found = false;
+                                for (int c = 0; c < pixLocalCounter; c++)
+                                {
+                                    if (pixLocalVolUnits[c] == tsdf_idx)
+                                    {
+                                        found = true; break;
+                                    }
+                                }
+                                if (!found)
+                                {
+                                    pixLocalVolUnits[pixLocalCounter++] = tsdf_idx;
+                                    if (pixLocalCounter >= pixCapacity)
+                                    {
+                                        return;
+                                    }
+                                }
+                            }
+                        }
+
+                // lock localAccessVolUnits somehow
+                for (int i = 0; i < pixLocalCounter; i++)
+                {
+                    Vec3i idx = pixLocalVolUnits[i];
+                    if (!ghm.insert(idx))
+                    {
+                        //return;
+                    }
+                }
+                // unlock
+            }
+        }
+    };
+
+    Rect dim(0, 0, depth.cols, depth.rows);
+    Size gsz(32, 32);
+    Size gg(divUp(dim.width, gsz.width), divUp(dim.height, gsz.height));
+
+    bool needReallocation = false;
+    auto allocateLambda = [&](const Range& r)
+    {
+
+        for (int yg = r.start; yg < r.end; yg++)
+        {
+            for (int xg = 0; xg < gg.width; xg++)
+            {
+                Rect gr(xg * gsz.width, yg * gsz.height, (xg + 1) * gsz.width, (yg + 1) * gsz.height);
+                gr = gr & dim;
+                Range xr(gr.tl().x, gr.br().x), yr(gr.tl().y, gr.br().y);
+
+                CustomHashSet ghm;
+
+                fillLocalAcessVolUnits(xr, yr, ghm);
+
+                if (ghm.last)
+                {
+                    std::lock_guard<std::recursive_mutex> al(mutex);
```

Comment:
Use OpenCV wrappers:
- cv::Mutex
- cv::AutoLock

Reply:
Why to use these OpenCV wrappers if they are just typedefs of STL sync primitives?

Reply:
> just typedefs

Because they don't in some configurations.
See #19985 or [Windows managed compilation](https://github.com/microsoft/STL/blob/main/stl/inc/mutex) (`_M_CEE` definition)